### PR TITLE
DDS-310 Remove built-in stateful writer

### PR DESCRIPTION
**Pull request by jrebelo**
This is a refactoring step which removes the built-in stateful writer and replaces its usage by the modified UserDefinedWriter.

### CODE REVIEW

File: dds/src/implementation/dds_impl/builtin_publisher.rs [hunk 1]
```diff
@@ -1,51 +1,64 @@
...
 };
 
 pub struct BuiltinPublisher {
     _qos: PublisherQos,
     _rtps_group: RtpsGroup,
     spdp_builtin_participant_writer: DdsShared<BuiltinStatelessWriter>,
-    sedp_builtin_topics_writer: DdsShared<BuiltinStatefulWriter>,
-    sedp_builtin_publications_writer: DdsShared<BuiltinStatefulWriter>,
-    sedp_builtin_subscriptions_writer: DdsShared<BuiltinStatefulWriter>,
+    sedp_builtin_topics_writer: DdsShared<UserDefinedDataWriter<RtpsStatefulWriter>>,
```
Comment (stkimmer) [inline dds/src/implementation/dds_impl/builtin_publisher.rs:51]: That the builtin is of type UserDefined is a little confusing

File: dds/src/domain/domain_participant_factory.rs [hunk 2]
```diff
@@ -138,7 +143,7 @@ fn configuration_try_from_str(configuration_json: &str) -> Result<DustDdsConfigu
 /// [`DomainParticipantFactory`] itself has no factory. It is a pre-existing singleton object that can be accessed by means of the
 /// [`DomainParticipantFactory::get_instance`] operation.
 pub struct DomainParticipantFactory {
-    participant_list: DdsRwLock<Vec<DcpsService>>,
+    participant_list: DdsRwLock<Vec<DdsShared<DcpsService>>>,
```
Comment (stkimmer) [inline dds/src/domain/domain_participant_factory.rs:146]: That the dcps service needs to be shared does sound dangerous. I was hoping the service uses other things, and is not used by other things.

File: dds/src/implementation/dds_impl/domain_participant_impl.rs [hunk 3]
```diff
@@ -903,78 +895,35 @@ impl DdsShared<DomainParticipantImpl> {
...
-                .sedp_builtin_subscriptions_writer()
-                .remove_matched_participant(participant_guid_prefix);
-            self.get_builtin_publisher()
-                .sedp_builtin_topics_writer()
-                .remove_matched_participant(participant_guid_prefix);
-        }
-
-        self.discovered_participant_remove(participant_handle);
-    }
+    // pub fn remove_discovered_participant(&self, participant_handle: InstanceHandle) {
```
Comment (stkimmer) [inline dds/src/implementation/dds_impl/domain_participant_impl.rs:898]: So it's not tested?

**Review by stkimmer — APPROVED** [review]

File: dds/src/implementation/dds_impl/domain_participant_impl.rs (hunk 3 above)
Comment (jrebelo) [reply dds/src/implementation/dds_impl/domain_participant_impl.rs:898]: It was used before by the ignore but not explicitly verified. So it isn't really tested

File: dds/src/domain/domain_participant_factory.rs (hunk 2 above)
Comment (jrebelo) [reply dds/src/domain/domain_participant_factory.rs:146]: I hope this is a temporary situation but it needed to be able to not have the condvar/sender inside the underlying objects so I accept it for now.

File: dds/src/implementation/dds_impl/builtin_publisher.rs (hunk 1 above)
Comment (jrebelo) [reply dds/src/implementation/dds_impl/builtin_publisher.rs]: I was planning to rename it later but I have done it now instead :)

**Pull request merged**